src/wallet/wallet.cpp

```diff
@@ -260,6 +260,66 @@ std::shared_ptr<CWallet> LoadWalletInternal(WalletContext& context, const std::s
         return nullptr;
     }
 }
+
+class FastWalletRescanFilter
+{
+public:
+    FastWalletRescanFilter(const CWallet& wallet) : m_wallet(wallet)
+    {
+        // fast rescanning via block filters is only supported by descriptor wallets right now
+        assert(!m_wallet.IsLegacy());
+
+        // create initial filter with scripts from both active and non-active ScriptPubKeyMans
+        auto active_spkms = m_wallet.GetActiveScriptPubKeyMans();
+        for (auto spkm : m_wallet.GetAllScriptPubKeyMans()) {
+            auto desc_spkm = dynamic_cast<DescriptorScriptPubKeyMan*>(spkm);
+            for (const auto& script_pub_key : desc_spkm->GetScriptPubKeys()) {
+                m_filter_set.emplace(script_pub_key.begin(), script_pub_key.end());
+            }
+            // save active ScriptPubKeyMans's end ranges for possible future filter updates
+            if (active_spkms.find(spkm) != active_spkms.end()) {
+                m_last_range_ends.emplace(desc_spkm, GetDescriptorRangeEnd(*desc_spkm));
+            }
+        }
+    }
+
+    void UpdateIfNeeded()
+    {
+        // repopulate filter with new scripts if top-up has happened since last iteration
+        for (auto [desc_spkm, last_range_end] : m_last_range_ends) {
+            int32_t current_range_end = GetDescriptorRangeEnd(*desc_spkm);
+            if (current_range_end > last_range_end) {
+                for (const auto& script_pub_key : desc_spkm->GetScriptPubKeys(last_range_end)) {
+                    m_filter_set.emplace(script_pub_key.begin(), script_pub_key.end());
+                }
+                m_last_range_ends.at(desc_spkm) = current_range_end;
+            }
+        }
+    }
+
+    std::optional<bool> MatchesBlock(const uint256& block_hash) const
+    {
+        return m_wallet.chain().blockFilterMatchesAny(BlockFilterType::BASIC, block_hash, m_filter_set);
+    }
+
+private:
+    const CWallet& m_wallet;
+    /** Map for keeping track of each active descriptor's last seen end range.
+      * This information is used to detect whether new addresses were derived
+      * (that is, if the current end range is larger than the saved end range)
+      * after processing a block and hence a filter set update is needed to
+      * take possible keypool top-ups into account.
+      */
+    std::map<const DescriptorScriptPubKeyMan*, int32_t> m_last_range_ends;
```

Comment:
I don't think that pointers are a good design choice for map keys. Would rather use the spkm id:
```suggestion
    std::map<uint256, int32_t> m_last_range_ends;
```

Reply:
Valid point, I struggled with the key type choice a lot. In an earlier version (before the PR was opened) I even used IDs at keys, but the drawback of this is that it leads to more code, as we (artificially?) need to convert between SPKM-Id and DescriptorScriptPubKey* pointers again and again, i.e. needing to call `CWallet::GetScriptPubKeyMan(const uint256& id)` (+casting the result to `DescriptorScriptPubKeyMan*`) at every iteration of the map and call `DescriptorScriptPubKeyMan::GetID()` before every insertion into the map. Considering that the result is always the same, it seems a bit excessive to recalculate the ID (involving creation of a string and hashing it, see `DescriptorScriptPubKeyMan::GetID()`) again and again. As long as we are sure that the different pointers are constant and unique, using them as keys should be okay? (At least, I couldn't find anything stating that it should be avoided, as long as we know what we are doing; e.g. https://stackoverflow.com/questions/25122932/pointers-as-keys-in-map-c-stl).

Any thoughts?